File: plane.cpp
#include <cmath>

#include "plane.hpp"

Plane::Plane(const Point &point, const Vector &normal, const Vector &color)
    : point(point), normal(normal.normalize()), color(color) {}

bool Plane::intersect(const Point &rayOrigin, const Vector &rayDirection, double &t) const
{
    // Calcula o numerador da fórmula da interseção do raio com o plano
    double numerator = point.difference(rayOrigin).dot(normal);

    // Calcula o denominador da fórmula da interseção do raio com o plano
    double denominator = rayDirection.dot(normal);

    // Verifica se o raio é paralelo ou quase paralelo ao plano
    if (std::abs(denominator) > 1e-6)
    {
        // Calcula o parâmetro 't' de interseção
        t = numerator / denominator;

        // Verifica se a interseção está na frente do observador
        if (t > 0.0)
        {
            return true;
        }
    }

    return false; // Não houve interseção
}

File: src/plane.py
from src.point import Point
from src.shape import Shape
from src.vector import Vector


class Plane(Shape):
    def __init__(self, point: Point, normal: Vector, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.point = point
        self.normal = normal

    def intersect(self, ray_origin: Point, ray_direction: Vector) -> tuple[bool, float]:
        den = self.normal.dot(ray_direction)
        if abs(den) > 0.0001:
            num = self.point.difference(ray_origin).dot(self.normal)
            t = num / den
            if t > 0:
                return True, t
        return False, 0

    def __str__(self):
        return f"Plane({self.point}, {self.normal})"
